### ISSUE TITLE
💅 noFloatingPromises does detect unawaited vitest expect(...).asyncFunction() calls

### ISSUE THREAD

**Issue by alexprudhomme** [biomejs/biome#6821]
### Environment information

```bash
biome rage --linter
CLI:
  Version:                      1.8.3
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           unset
  JS_RUNTIME_NAME:              unset
  NODE_PACKAGE_MANAGER:         unset

Biome Configuration:
  Error:                        Found an unknown key `root`.
  Error:                        extends has an incorrect type, expected an array, but received a string.
  Error:                        Found an unknown key `includes`.
  Error:                        Found an unknown key `noFloatingPromises`.
  Error:                        Found an unknown key `assist`.
  Status:                       Loaded with errors
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    false
  VCS disabled:                 true

Linter:
  JavaScript enabled:           true
  JSON enabled:                 true
  CSS enabled:                  false
  Recommended:                  true
  All:                          false
  Enabled rules:
  performance/noDelete
  suspicious/noCatchAssign
  suspicious/noUnsafeNegation
  complexity/useLiteralKeys
  style/useImportType
  complexity/noMultipleSpacesInRegularExpressionLiterals
  a11y/useValidLang
  complexity/noUselessEmptyExport
  suspicious/useNamespaceKeyword
  suspicious/useValidTypeof
  a11y/useValidAriaRole
  correctness/noConstantCondition
  a11y/useAriaActivedescendantWithTabindex
  suspicious/noAssignInExpressions
  style/useDefaultParameterLast
  complexity/noEmptyTypeParameters
  correctness/noConstructorReturn
  style/useSelfClosingElements
  suspicious/noDuplicateParameters
  style/useTemplate
  correctness/noUnusedLabels
  complexity/noUselessTernary
  correctness/noUnreachableSuper
  suspicious/noCompareNegZero
  suspicious/noExplicitAny
  correctness/noSwitchDeclarations
  a11y/noAutofocus
  correctness/noUnsafeOptionalChaining
  correctness/noConstAssign
  suspicious/noControlCharactersInRegex
  complexity/noUselessTypeConstraint
  style/noVar
  suspicious/noDoubleEquals
  suspicious/noRedundantUseStrict
  style/useLiteralEnumMembers
  suspicious/noGlobalIsNan
  suspicious/noEmptyInterface
  suspicious/noConstEnum
  suspicious/noMisleadingCharacterClass
  correctness/noPrecisionLoss
  suspicious/noRedeclare
  correctness/noStringCaseMismatch
  correctness/noSetterReturn
  correctness/noInvalidConstructorSuper
  suspicious/noImplicitAnyLet
  suspicious/noFallthroughSwitchClause
  suspicious/noUnsafeDeclarationMerging
  correctness/noUnreachable
  a11y/useKeyWithClickEvents
  suspicious/noDuplicateObjectKeys
  complexity/noUselessThisAlias
  complexity/noThisInStatic
  complexity/useOptionalChain
  correctness/noInnerDeclarations
  style/noParameterAssign
  suspicious/noDuplicateCase
  a11y/useValidAnchor
  complexity/useRegexLiterals
  correctness/noSelfAssign
  style/noUselessElse
  style/useShorthandFunctionType
  suspicious/noShadowRestrictedNames
  a11y/useMediaCaption
  complexity/noUselessLabel
  complexity/noUselessCatch
  correctness/noUnsafeFinally
  a11y/useAriaPropsForRole
  correctness/noNonoctalDecimalEscape
  style/useEnumInitializers
  a11y/useHtmlLang
  suspicious/noDuplicateTestHooks
  complexity/noStaticOnlyClass
  style/useWhile
  complexity/useArrowFunction
  style/noInferrableTypes
  a11y/noNoninteractiveTabindex
  complexity/useSimpleNumberKeys
  correctness/useYield
  a11y/noInteractiveElementToNoninteractiveRole
  style/useNumericLiterals
  correctness/noUnnecessaryContinue
  suspicious/noApproximativeNumericConstant
  suspicious/noImportAssign
  suspicious/noLabelVar
  correctness/noGlobalObjectCalls
  suspicious/useDefaultSwitchClauseLast
  a11y/useAltText
  correctness/noEmptyCharacterClassInRegex
  suspicious/noSuspiciousSemicolonInJsx
  suspicious/noSparseArray
  a11y/useIframeTitle
  complexity/noBannedTypes
  a11y/noSvgWithoutTitle
  correctness/noVoidElementsWithChildren
  style/useAsConstAssertion
  correctness/useJsxKeyInIterable
  style/useExportType
  complexity/noUselessLoneBlockStatements
  style/noArguments
  a11y/useValidAriaValues
  suspicious/noDebugger
  suspicious/noMisleadingInstantiator
  suspicious/noPrototypeBuiltins
  suspicious/noGlobalAssign
  suspicious/noCommentText
  suspicious/noDuplicateJsxProps
  suspicious/noThenProperty
  a11y/noPositiveTabindex
  correctness/noEmptyPattern
  complexity/noExcessiveNestedTestSuites
  security/noDangerouslySetInnerHtmlWithChildren
  a11y/useKeyWithMouseEvents
  suspicious/noExtraNonNullAssertion
  suspicious/useGetterReturn
  correctness/noRenderReturnValue
  correctness/useExhaustiveDependencies
  security/noGlobalEval
  style/useConst
  a11y/noRedundantRoles
  complexity/useFlatMap
  correctness/useIsNan
  suspicious/noGlobalIsFinite
  suspicious/noSelfCompare
  suspicious/noAsyncPromiseExecutor
  security/noDangerouslySetInnerHtml
  style/useNodejsImportProtocol
  a11y/noDistractingElements
  suspicious/noArrayIndexKey
  complexity/noWith
  suspicious/noDuplicateClassMembers
  complexity/noExtraBooleanCast
  performance/noAccumulatingSpread
  a11y/useValidAriaProps
  a11y/noRedundantAlt
  correctness/noChildrenProp
  suspicious/noConfusingLabels
  suspicious/noConfusingVoidType
  suspicious/noFocusedTests
  a11y/useButtonType
  a11y/noAriaUnsupportedElements
  correctness/noFlatMapIdentity
  a11y/noBlankTarget
  a11y/useHeadingContent
  correctness/useValidForDirection
  correctness/noVoidTypeReturn
  correctness/noInvalidUseBeforeDeclaration
  a11y/noAriaHiddenOnFocusable
  a11y/useAnchorContent
  complexity/noUselessRename
  correctness/noInvalidNewBuiltin
  style/useNumberNamespace
  complexity/noUselessConstructor
  a11y/noAccessKey
  style/useExponentiationOperator
  style/noUnusedTemplateLiteral
  complexity/noUselessSwitchCase
  style/useSingleVarDeclarator
  suspicious/noExportsInTest
  a11y/noNoninteractiveElementToInteractiveRole
  style/noCommaOperator
  suspicious/useIsArray
  a11y/noHeaderScope
  complexity/noUselessFragments
  suspicious/noMisrefactoredShorthandAssign
  complexity/noForEach
  suspicious/noClassAssign
  suspicious/noFunctionAssign
```

### Rule name

noFloatingPromises

### Playground link

https://biomejs.dev/playground/?lintRules=all&code=ZgB1AG4AYwB0AGkAbwBuACAAZQB4AHAAZQBjAHQAKABwAHIAbwBtAGkAcwBlADoAIABQAHIAbwBtAGkAcwBlADwAdQBuAGsAbgBvAHcAbgA%2BACkAIAB7AAoAIAAgAHIAZQB0AHUAcgBuACAAewAKACAAIAAgACAAdABvAEIAZQBWAGkAcwBpAGIAbABlADoAIABhAHMAeQBuAGMAIAAoACkAIAA9AD4AIAB7AAoAIAAgACAAIAAgACAAYQB3AGEAaQB0ACAAcAByAG8AbQBpAHMAZQA7AAoAIAAgACAAIAB9ACwACgAgACAAfQA7AAoAfQAKAAoAYQBzAHkAbgBjACAAZgB1AG4AYwB0AGkAbwBuACAAZwBlAHQAVgBhAGwAdQBlACgAKQAgAHsACgAgACAAcgBlAHQAdQByAG4AIAA0ADIAOwAKAH0ACgAKAGUAeABwAGUAYwB0ACgAZwBlAHQAVgBhAGwAdQBlACgAKQApAC4AdABvAEIAZQBWAGkAcwBpAGIAbABlACgAKQA7AAoAYQB3AGEAaQB0ACAAZQB4AHAAZQBjAHQAKABnAGUAdABWAGEAbAB1AGUAKAApACkALgB0AG8AQgBlAFYAaQBzAGkAYgBsAGUAKAApADsACgAKACAA

### Expected result

It should throw an error. 

Something like this does not throw the noFloatingPromises error even though isVisible is an async function.

```
function expect(promise: Promise<unknown>) {
  return {
    toBeVisible: async () => {
      await promise;
    },
  };
}

expect(getValue()).toBeVisible();
await expect(getValue()).toBeVisible();

```
 

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by ematipico** [comment]
@alexprudhomme your `biome rage` was using an old version of Biome. I assume it was a mistake, so can you please fix it?

**Comment by alexprudhomme** [comment]
@ematipico done, we are using 2.0.6

**Comment by alexprudhomme** [comment]
Created a repro : https://github.com/alexprudhomme/biome-repro-1752344509915

This happens because biome is not catching what Promisify does in vitest : https://github.com/vitest-dev/vitest/blob/7b489959a0a734727139acee31b46cf354868127/packages/browser/matchers.d.ts#L9

This might be impossible to fix so feel free to close  ! 🤷

**Comment by arendjr** [comment]
Thanks, this is a useful scenario to support! We don't currently support mapped types yet, but it's something we want to support eventually.
